### PR TITLE
HW18 gamzinova

### PR DESCRIPTION
**Pull request by gamzinova**
Good evening,
Please find below my homework, and thanks a lot for your help. I have solved the issue with circular imports.

It was a very difficult but on the other side interesting homework.  I am working with server testing, so thank you for the opportunity to try web testing.

Best,
Anna

### CODE REVIEW

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/checkout_page.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    def error_message_checkout(self):
+        return self.is_displayed(locators.error_message_checkout)
+
+    def cancel_checkout_button(self):
+        """
+         A function that forwards the user to the Main page after clicking the cancel button on the checkout page
+        """
+        self.click(locators.cancel_button)
+        return self
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/checkout_page.py:48]: it is better remove return here

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/finish_page.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class FinishPage(BasePage):
+    def __init__(self, driver):
+        super().__init__(driver)
+
+    def is_thanks_header(self):
+        return self.is_displayed(locators.thanks_header)
+
+    def label_picture(self):
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/finish_page.py:12]: if function return bool it's name should start with is

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/item_page.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.cart_page import CartPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.web_ui.base_page import BasePage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities import locators
+
+
+class ItemPage(BasePage):
+    def __init__(self, driver):
+        super().__init__(driver)
+
+    def picture_item(self):
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/item_page.py:10]: is_pi....

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/item_page.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        return self
+
+    def is_add_to_cart_button(self):
+        return self.is_displayed(locators.add_to_cart_button)
+
+    def remove_from_cart_item_page(self):
+        self.click(locators.remove_the_item_item_page)
+        return self
+
+    def remove_item(self):
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/item_page.py:28]: is_....

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/main_page.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        self.click(locators.cart_button)
+        return CartPage(self._driver)
+
+    def add_to_cart_click(self):
+        return self.click(locators.first_item_add_to_cart)
+
+    def remove_from_cart(self):
+        return self.click(locators.first_item_remove)
+
+    def remove_item(self):
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/main_page.py:39]: is_...

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/overview_page.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.finish_page import FinishPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.web_ui.base_page import BasePage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities import locators
+
+
+class OverviewPage(BasePage):
+    def __init__(self, driver):
+        super().__init__(driver)
+
+    def payment_information(self):
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/overview_page.py:10]: is_...

File: Anna_Gamzinova/HW_18_framework_from_scratch/utilities/driver_factory.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class DriverFactory:
+    CHROME = 1
+    FIREFOX = 2
+    EDGE = 3
+
+    @staticmethod
+    def create_driver(driver_id: int):
+        if int(driver_id) == 1:
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/utilities/driver_factory.py:15]: please fix if int(driver_id) == DriverFactory.CHROME:

File: Anna_Gamzinova/HW_18_framework_from_scratch/utilities/locators.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+css_first_item_page = (By.CSS_SELECTOR, '[alt="Sauce Labs Backpack"]')
+css_first_item_add_to_cart = (By.CSS_SELECTOR, '[name="add-to-cart-sauce-labs-backpack"]')
+
+# item page:
+css_back_to_products_button = (By.CSS_SELECTOR, '#back-to-products')
+css_item_picture = (By.CSS_SELECTOR, '[alt="Sauce Labs Backpack"]')
+css_add_to_cart_button = (By.CSS_SELECTOR, '#add-to-cart-sauce-labs-backpack')
+css_remove_from_the_cart_button = (By.CSS_SELECTOR, '#remove-sauce-labs-backpack')
+css_continue_shopping_button = (By.CSS_SELECTOR, '#continue-shopping')
+css_checkout_button = (By.CSS_SELECTOR, '#checkout')
```
Comment (bohdandovbsyh) [inline Anna_Gamzinova/HW_18_framework_from_scratch/utilities/locators.py:89]: it is not good approach to store all locators i one file

File: Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
+import pytest
+
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.main_page import MainPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.login_page import LoginPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.config_parser import ReadConfig
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.driver_factory import DriverFactory
+
+
+@pytest.fixture(scope='session')
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py]: fixed

File: Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+@pytest.fixture()
+def open_login_page(create_driver):
+    return LoginPage(create_driver)
+
+
+@pytest.fixture()
+def open_main_page(open_login_page):
+    login_page = open_login_page
+    main_page = login_page.login(ReadConfig.get_user_name(), ReadConfig.get_user_password())
+    return main_page
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py]: done

File: Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+
+@pytest.fixture()
+def init_main_page(create_driver):
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/conftest.py:32]: added an explanation to the file

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.checkout_page import CheckoutPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.web_ui.base_page import BasePage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities import locators
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py]: moved to the folder locators

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        super().__init__(driver)
+
+    def cart_item(self):
+        return self.is_displayed(locators.cart_button)
+
+    def first_item_in_the_cart(self):
+        return self.is_displayed(locators.cart_item)
+
+    def remove_item_from_the_cart(self):
+        return self.click(locators.remove_from_the_cart_button)
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py]: fixed

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
+from Anna_Gamzinova.HW_18_framework_from_scratch.page_objects.checkout_page import CheckoutPage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities.web_ui.base_page import BasePage
+from Anna_Gamzinova.HW_18_framework_from_scratch.utilities import locators
+
+
+class CartPage(BasePage):
+    def __init__(self, driver):
+        super().__init__(driver)
+
+    def cart_item(self):
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py]: done

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+    def menu_click(self):
+        self.click(locators.menu_button)
+        return self
+
+    def logout_click(self):
+        self.click(locators.logout_button)
+        return self
+
+    def checkout_button_click(self):
```
[the remark this answers is not in the document]
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/cart_page.py]: added for all pageObjects

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/checkout_page.py (hunk 1 above)
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/checkout_page.py]: done

File: Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/finish_page.py (hunk 2 above)
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/page_objects/finish_page.py]: fixed

File: Anna_Gamzinova/HW_18_framework_from_scratch/utilities/driver_factory.py (hunk 7 above)
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/utilities/driver_factory.py]: @bohdandovbsyh done

File: Anna_Gamzinova/HW_18_framework_from_scratch/utilities/locators.py (hunk 8 above)
Comment (gamzinova) [reply Anna_Gamzinova/HW_18_framework_from_scratch/utilities/locators.py]: @bohdandovbsyh have separated locators into two fils

**Pull request merged**